hal-core/src/interrupt.rs

```diff
@@ -0,0 +1,125 @@
+/// An interrupt controller for a platform.
+pub trait Control {
+    type Error;
+    /// The type of this platform's interrupt vector.
+    ///
+    /// TODO(eliza): This is *probably* always a `u8`, do we really need to have
+    /// this?
+    type Vector;
+
+    /// Disable interrupts.
+    unsafe fn disable_irq(&mut self);
+
+    /// Enable interrupts.
+    unsafe fn enable_irq(&mut self);
+
+    /// Returns `true` if interrupts are enabled.
+    fn is_enabled(&self) -> bool;
+
+    /// Register an interrupt handler to service the provided interrupt.
+    ///
+    /// The interrupt controller is assumed to generate an actual ISR that calls
+    /// the handler; that ISR will construct an interrupt context for the
+    /// handler. The ISR is responsible for entering a critical section while
+    /// the handler is active.
+    fn register_handler<I>(
+        &mut self,
+        irq: &I,
+        handler: I::Handler,
+    ) -> Result<(), RegistrationError>
+    where
+        I: Interrupt<Ctrl = Self>;
+
+    /// Enter a critical section, returning a guard.
+    fn enter_critical(&mut self) -> CriticalGuard<'_, Self> {
```

Comment:
yay this is nice

Reply:
it could probably also just be called `critical`, but i was thinking about having a version taking a closure. there’s not really a compelling reason to have both though, and the drop guard is a more flexible api